File: src/test/java/org/crafter_unit_tests/ChunkBitShiftTest.java
/*
 Crafter - A blocky game (engine) written in Java with LWJGL.
 Copyright (C) 2023  jordan4ibanez

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.crafter_unit_tests;

import org.crafter.engine.world.chunk.Chunk;
import org.joml.Random;
import org.junit.jupiter.api.Test;

import java.util.Date;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ChunkBitShiftTest {

    private static final int testAmount = 65_535;

    @Test
    public void testChunkBlockID() {

        Random random = new Random((int) (new Date().getTime()/1000));

        for (int i = 0; i < testAmount; i++) {
            final int chosenID = random.nextInt(65_535);
            final int testBlock = Chunk.setBlockID(0, chosenID);
            final int gottenID = Chunk.getBlockID(testBlock);
            assertEquals(chosenID, gottenID);
        }
    }

    @Test
    public void testChunkBlockLight() {

        Random random = new Random((int) (new Date().getTime()/1000));

        for (int i = 0; i < testAmount; i++) {
            final int chosenLight = random.nextInt(15);
            final int testBlock = Chunk.setBlockLight(0, chosenLight);
            final int gottenLight = Chunk.getBlockLight(testBlock);
            assertEquals(chosenLight, gottenLight);
        }
    }

    @Test
    public void testChunkBlockState() {

        Random random = new Random((int) (new Date().getTime()/1000));

        for (int i = 0; i < testAmount; i++) {
            final int chosenState = random.nextInt(16);
            final int testBlock = Chunk.setBlockState(0, chosenState);
            final int gottenState = Chunk.getBlockState(testBlock);
            assertEquals(chosenState, gottenState);
        }
    }

    /**
     * Now go crazy. Reuse a block over and over and see if it breaks.
     */
    @Test
    public void testAllBitManipulation() {

        Random random = new Random((int) (new Date().getTime()/1000));

        int testBlock = 0;

        for (int x = 0; x < testAmount; x++) {

            final int chosenID = random.nextInt(65_535);
            final int chosenLight = random.nextInt(15);
            final int chosenState = random.nextInt(15);

            testBlock = Chunk.setBlockID(testBlock, chosenID);

            assertEquals(chosenID, Chunk.getBlockID(testBlock));

            testBlock = Chunk.setBlockLight(testBlock, chosenLight);

            assertEquals(chosenID, Chunk.getBlockID(testBlock));
            assertEquals(chosenLight, Chunk.getBlockLight(testBlock));

            testBlock = Chunk.setBlockState(testBlock, chosenState);

            assertEquals(chosenID, Chunk.getBlockID(testBlock));
            assertEquals(chosenLight, Chunk.getBlockLight(testBlock));
            assertEquals(chosenState, Chunk.getBlockState(testBlock));

            // Now we're going in reverse order to double-check

            testBlock = Chunk.setBlockLight(testBlock, chosenLight);

            assertEquals(chosenID, Chunk.getBlockID(testBlock));
            assertEquals(chosenLight, Chunk.getBlockLight(testBlock));
            assertEquals(chosenState, Chunk.getBlockState(testBlock));

            testBlock = Chunk.setBlockID(testBlock, chosenID);

            assertEquals(chosenID, Chunk.getBlockID(testBlock));
            assertEquals(chosenLight, Chunk.getBlockLight(testBlock));
            assertEquals(chosenState, Chunk.getBlockState(testBlock));
        }

    }
}

File: src/test/kotlin/org/crafter_unit_tests/ChunkBitShiftTest.kt
package org.crafter_unit_tests

import org.crafter.engine.world.chunk.Chunk
import org.joml.Random
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.*

class ChunkBitShiftTest {
    @Test
    fun testChunkBlockID() {
        val testChunk = Chunk(0, 0)
        val random = Random((Date().time / 1000).toInt().toLong())
        for (i in 0 until testAmount) {
            val chosenID = random.nextInt(65535)
            val testBlock = testChunk.setBlockID(0, chosenID)
            val gottenID = testChunk.getBlockID(testBlock)
            Assertions.assertEquals(chosenID, gottenID)
        }
    }

    @Test
    fun testChunkBlockLight() {
        val testChunk = Chunk(0, 0)
        val random = Random((Date().time / 1000).toInt().toLong())
        for (i in 0 until testAmount) {
            val chosenLight = random.nextInt(15)
            val testBlock = testChunk.setBlockLight(0, chosenLight)
            val gottenLight = testChunk.getBlockLight(testBlock)
            Assertions.assertEquals(chosenLight, gottenLight)
        }
    }

    @Test
    fun testChunkBlockState() {
        val testChunk = Chunk(0, 0)
        val random = Random((Date().time / 1000).toInt().toLong())
        for (i in 0 until testAmount) {
            val chosenState = random.nextInt(16)
            val testBlock = testChunk.setBlockState(0, chosenState)
            val gottenState = testChunk.getBlockState(testBlock)
            Assertions.assertEquals(chosenState, gottenState)
        }
    }

    /**
     * Now go crazy. Reuse a block over and over and see if it breaks.
     */
    @Test
    fun testAllBitManipulation() {
        val testChunk = Chunk(0, 0)
        val random = Random((Date().time / 1000).toInt().toLong())
        var testBlock = 0
        for (x in 0 until testAmount) {
            val chosenID = random.nextInt(65535)
            val chosenLight = random.nextInt(15)
            val chosenState = random.nextInt(15)
            testBlock = testChunk.setBlockID(testBlock, chosenID)
            Assertions.assertEquals(chosenID, testChunk.getBlockID(testBlock))
            testBlock = testChunk.setBlockLight(testBlock, chosenLight)
            Assertions.assertEquals(chosenID, testChunk.getBlockID(testBlock))
            Assertions.assertEquals(chosenLight, testChunk.getBlockLight(testBlock))
            testBlock = testChunk.setBlockState(testBlock, chosenState)
            Assertions.assertEquals(chosenID, testChunk.getBlockID(testBlock))
            Assertions.assertEquals(chosenLight, testChunk.getBlockLight(testBlock))
            Assertions.assertEquals(chosenState, testChunk.getBlockState(testBlock))

            // Now we're going in reverse order to double-check
            testBlock = testChunk.setBlockLight(testBlock, chosenLight)
            Assertions.assertEquals(chosenID, testChunk.getBlockID(testBlock))
            Assertions.assertEquals(chosenLight, testChunk.getBlockLight(testBlock))
            Assertions.assertEquals(chosenState, testChunk.getBlockState(testBlock))
            testBlock = testChunk.setBlockID(testBlock, chosenID)
            Assertions.assertEquals(chosenID, testChunk.getBlockID(testBlock))
            Assertions.assertEquals(chosenLight, testChunk.getBlockLight(testBlock))
            Assertions.assertEquals(chosenState, testChunk.getBlockState(testBlock))
        }
    }

    companion object {
        private const val testAmount = 65535
    }
}
